examples/admission_controller.rs

```diff
@@ -0,0 +1,93 @@
+    tracing_subscriber::fmt::init();
+
+    let routes = warp::path("mutate")
+        .and(warp::body::json())
+        .and_then(mutate)
+        .with(warp::trace::request());
+
+
+    // For admission controllers running behind a service, you must generate a
+    // certificate for the service and encode the CA in the
+    // MutatingWebhookConfiguration, then terminate TLS here. For controllers
+    // running outside of the cluster, while still necessary, it's fine to
+    // handle TLS upstream. See:
+    // https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/#tls-certificates
+
+    // warp::serve(warp::post().and(routes))
+    //     .tls()
+    //     .cert_path("admission-controller-tls.crt")
+    //     .key_path("admission-controller-tls.key")
+    //     .run(([0, 0, 0, 0], 8443))
+    //     .await;
+
+    warp::serve(warp::post().and(routes))
+        .run(([0, 0, 0, 0], 8080))
+        .await;
+}
+
+async fn mutate(body: AdmissionReview<DynamicObject>) -> Result<impl Reply, Infallible> {
+    let mut res: AdmissionResponse = Default::default();
+
+    let req: AdmissionRequest<DynamicObject> = match body.try_into() {
+        Ok(req) => req,
+        Err(err) => {
+            res.allowed = false;
+            res.result.message = Some(err.to_string());
+
+            error!("invalid request: {}", err);
+            return Ok(reply::json(&res.into_review()));
+        }
+    };
+
+    res.uid = req.uid.clone();
+    // Default to allowed if we don't find anything.
+    res.allowed = true;
+
+    if let Some(obj) = req.object {
+        info!(
+            "got request id {} to {:?} resource {}",
+            req.uid.clone(),
+            req.operation,
+            obj.metadata.name.unwrap_or("unknown".to_owned())
+        );
+
+        let mut patches: Vec<json_patch::PatchOperation> = Vec::new();
+        let labels = obj.metadata.labels.unwrap_or_default();
+
+        // If the resource contains a label named "test_reject", it will be
+        // forbidden from creating.
+        if labels.contains_key("test_reject") {
+            res.allowed = false;
+            res.result.message = Some("Resource contained 'test_reject' label.".to_owned());
+
+            return Ok(reply::json(&res.into_review()));
+        }
+
+        // If the resource doesn't contain "test_modify", we add it to the
+        // resource.
+        if !labels.contains_key("test_modify") {
+            patches.push(json_patch::PatchOperation::Add(json_patch::AddOperation {
+                path: "/metadata/labels/test_modify".to_owned(),
+                value: serde_json::Value::String("modified-by-admission-controller".to_owned()),
+            }))
+        }
+
+        if !patches.is_empty() {
+            res.patch = serde_json::to_vec(&json_patch::Patch(patches)).ok();
+            if res.patch.is_some() {
+                res.patch_type = Some(PatchType::JsonPatch);
+            }
```

Comment:
This does feel a bit awkward when we already have a patch type that behaves differently. I think a builder on `AdmissionResponse` here that simply takes the json patch, and internally sets the type would be nicer.